sdk/appcenter/src/main/java/com/microsoft/appcenter/http/DefaultHttpClient.java

```diff
@@ -134,26 +151,55 @@ private static String doHttpCall(String urlString, String method, Map<String, St
             /* Configure connection timeouts. */
             urlConnection.setConnectTimeout(CONNECT_TIMEOUT);
             urlConnection.setReadTimeout(READ_TIMEOUT);
+
+            /* Build payload now if POST. */
             urlConnection.setRequestMethod(method);
+            String payload = null;
+            byte[] binaryPayload = null;
+            boolean shouldCompress = false;
+            boolean isPost = method.equals(METHOD_POST);
+            if (isPost && callTemplate != null) {
+
+                /* Get bytes, check if large enough to compress. */
+                payload = callTemplate.buildRequestBody();
+                binaryPayload = payload.getBytes(CHARSET_NAME);
+                shouldCompress = binaryPayload.length >= MIN_GZIP_LENGTH;
+
+                /* If no content type specified, assume json. */
+                if (!headers.containsKey(CONTENT_TYPE_KEY)) {
+                    headers.put(CONTENT_TYPE_KEY, CONTENT_TYPE_VALUE);
+                }
+            }
 
-            /* Set headers. */
-            urlConnection.setRequestProperty(CONTENT_TYPE_KEY, CONTENT_TYPE_VALUE);
+            /* If about to compress, add corresponding header. */
+            if (shouldCompress) {
+                headers.put(CONTENT_ENCODING_KEY, CONTENT_ENCODING_VALUE);
+            }
+
+            /* Send headers. */
             for (Map.Entry<String, String> header : headers.entrySet()) {
                 urlConnection.setRequestProperty(header.getKey(), header.getValue());
             }
 
-            /* Before send. */
+            /* Call back before the payload is sent. */
             if (callTemplate != null) {
                 callTemplate.onBeforeCalling(url, headers);
             }
 
-            /* Build payload. */
-            if (method.equals(METHOD_POST) && callTemplate != null) {
-                String payload = callTemplate.buildRequestBody();
+            /* Send payload. */
+            if (binaryPayload != null) {
+
+                /* Compress payload if large enough to be worth it. */
                 AppCenterLog.verbose(LOG_TAG, payload);
+                if (shouldCompress) {
+                    ByteArrayOutputStream gzipBuffer = new ByteArrayOutputStream(binaryPayload.length);
+                    GZIPOutputStream gzipStream = new GZIPOutputStream(gzipBuffer);
```

Comment:
You don't need `try` - `catch` to close output stream in error case?

Reply:
**it never fails** -> How do you guarantee?

Reply:
It should be okay since ByteArrayOutputStream.close doesn't do anything.

Reply:
Its in the source code of the 2 classes used in that block, it's all in memory. Gzip will only fail if wrapping a stream that fails (like a network stream), but here we wrap in memory.